[Mips] Guard emission of CFI instructions by `MF.needsFrameMoves()`

s-barannikov opened the pull request:
Don't emit CFI instructions that will be skipped by AsmPrinter. Also, use a helper class to simplify emission.

As a side effect, this seems to unblock delay slot filler optimization on some tests, though it is not obvious to me whether the transformations are legitimate.

Similar to #135845 and #136060.

s-barannikov commented:
I'll commit the NFC change separately to simplify the review.

s-barannikov commented:
Extracted the NFC part into #136242. I'll let it hang for a couple of days in case there are active reviewers.

yingopq commented on llvm/test/CodeGen/Mips/fpbr.ll:
```diff
@@ -20,7 +20,7 @@ entry:
 ; GPR:           not      $[[GPRCC]], $[[GPRCC]]
 ; 32-GPR:        bnez     $[[GPRCC]], $BB0_2
-; 64-GPR:        bnezc    $[[GPRCC]], .LBB0_2
+; 64-GPR:        bnez     $[[GPRCC]], .LBB0_2
```
I think it is better to add an explanation.

s-barannikov replied:
Sure, I'll investigate.

s-barannikov replied:
It appears it is the same delay slot filler optimization kicking in. The test doesn't show it, but the `bnez` instruction now has a store instruction in the delay slot. Before this patch, there was no delay slot candidate and the branch was transformed to the compact variant `bnezc` as explained in this comment:

```
    // For microMIPS if instruction is BEQ or BNE with one ZERO register, then
    // instead of adding NOP replace this instruction with the corresponding
    // compact branch instruction, i.e. BEQZC or BNEZC. Additionally
    // PseudoReturn and PseudoIndirectBranch are expanded to JR_MM, so they can
    // be replaced with JRC16_MM.

    // For MIPSR6 attempt to produce the corresponding compact (no delay slot)
    // form of the CTI. For indirect jumps this will not require inserting a
    // NOP and for branches will hopefully avoid requiring a NOP.
```

s-barannikov commented:
I'll try to get back to it next week, thanks for reminding

s-barannikov committed:
[Mips] Guard emission of CFI instructions by MF.needsFrameMoves()

Don't emit CFI instructions when they will be skipped by AsmPrinter.
Also, use a helper class to simplify emission.

As a side effect, this seems to have unblocked delay slot filler
optimization on some tests, though it is not obvious to me whether
the transformations are legitimate.

Similar to #135845 and #136060.

wzssyqa commented:
I think that we need a real test for this changes: which codes or command lines will produce different output?
